manage-todos-with-realtime-agent/realtime_over_websockets/main.py

```diff
@@ -0,0 +1,183 @@
+    )
+    data = get_data()
+    if len(data) > 0:
+        system_message += f"\n\nTodos from previous session:\n{format_todo_str(data)}"
+    else:
+        system_message += "\n\nNo todos from previous session."
+
+    realtime_agent = RealtimeAgent(
+        name="Weather Bot",
+        system_message=system_message,
+        llm_config=realtime_llm_config,
+        audio_adapter=audio_adapter,
+        logger=logger,
+    )
+
+    @realtime_agent.register_realtime_function(  # type: ignore [misc]
+        name="add_todo", description="Add a todo"
+    )
+    def add_todo(
+        id: Annotated[int, "task id"],
+        task: Annotated[str, "task description"],
+        status: Annotated[str, "task status, either done or open"],
+    ):
+        # if exist, load the data
+        # if not exist, create empty list
+        data = get_data()
+        data.append(
+            {
+                "id": id,
+                "task": task,
+                "status": status,
+            }
+        )
+        save_data(data)
+        return f"Todo with id {id} added.\n" + format_todo_str(data)
+
+    @realtime_agent.register_realtime_function(  # type: ignore [misc]
+        name="modify_todo", description="Modify a todo"
+    )
+    def modify_todo(
+        id: Annotated[int, "task id"],
+        task: Annotated[str, "task description"],
+        status: Annotated[str, "task status, either done or open"],
+    ):
+        data = get_data()
+        is_todo_found = False
+        for todo in data:
+            if todo["id"] == id:
+                todo["task"] = task
+                todo["status"] = status
+                is_todo_found = True
+                break
+
+        if not is_todo_found:
+            msg = f"Todo with id {id} not found."
+            add_msg = add_todo(id, task, status)
+            return msg + "\n" + add_msg
+
+        save_data(data)
+        return f"Todo with id {id} updated.\n" + format_todo_str(data)
+
+    @realtime_agent.register_realtime_function(  # type: ignore [misc]
+        name="delete_todo", description="Delete a todo"
+    )
+    def delete_todo(
+        id: Annotated[int, "task id"],
+        task: Annotated[str, "task description"],
+        status: Annotated[str, "task status, either done or open"],
+    ):
+        logger.info(f"Deleting todo with id {id}, task {task}")
+        data = get_data()
+
+        is_todo_found = False
+        new_data = []
+        for todo in data:
+            if todo["id"] == id:
+                is_todo_found = True
+            else:
+                new_data.append(todo)
+
+        data = new_data
+        save_data(data)
+        if is_todo_found:
+            msg = f"Todo with id {id} deleted.\n" + format_todo_str(data)
+        else:
+            msg = f"Todo with id {id} not found. Nothing changed.\n" + format_todo_str(
+                data
+            )
+        logger.info(f"Msg from delete: {msg}")
+        return msg
```

Comment:
Use the delete_todo from todo_utils

```suggestion
    @realtime_agent.register_realtime_function(  # type: ignore [misc]
        name="delete_todo", description="Delete a todo"
    )(delete_todo)
```